dali/operators/reader/numpy_reader_op.cc

```diff
@@ -14,15 +14,126 @@
 
 #include <string>
 
+#include "dali/pipeline/data/views.h"
 #include "dali/kernels/common/transpose.h"
 #include "dali/core/static_switch.h"
 #include "dali/operators/reader/numpy_reader_op.h"
 
 namespace dali {
 
-#define NUMPY_ALLOWED_TYPES \
-  (bool, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t, float, float16, \
-  double)
+void NumpyReader::SanitizeSliceArgs(TensorListShape<>& anchors, const TensorListShape<>& shapes) {
+  // if shape is set but anchor isn't, set anchor to 0
+  if (!shapes.empty() && anchors.empty()) {
+    TensorShape<> ts;
+    ts.resize(shapes.sample_dim());
+    for (int d = 0; d < shapes.sample_dim(); d++) ts[d] = 0;
+    anchors = uniform_list_shape(batch_size_, ts);
+  }
+}
+
+void NumpyReader::GetStaticSliceArg(TensorListShape<>& tls, const char *name) {
+  if (spec_.HasArgument(name)) {
+    tls = uniform_list_shape(batch_size_, spec_.GetRepeatedArgument<int>(name));
+  } else {
+    // create an list of empty shapes
+    tls = {};
+  }
+}
+
+void NumpyReader::GetDynamicSliceArg(TensorListShape<>& tls,
+                                     ArgumentWorkspace &ws,
+                                     const char *name) {
+  if (spec_.HasTensorArgument(name)) {
+    auto &t = ws.ArgumentInput(name);
+    DALI_ENFORCE(static_cast<int>(t.size()) == batch_size_
+                 && t.shape().sample_dim() == 1
+                 && is_uniform(t.shape()),
+                 "Shape must be a list of 1D tensors of equal length");
+    auto tlv = view<const int>(t);
+    tls.resize(batch_size_, t.shape()[0][0]);
+    TensorShape<> ts;
+    ts.resize(t.shape()[0][0]);
+    for (int i = 0; i < batch_size_; i++) {
+      auto tv = tlv[i];
+      for (int j = 0; j < tv.shape[0]; j++) {
+        ts[j] = tv.data[j];
+      }
+      tls.set_tensor_shape(i, ts);
+    }
+  } else {
+    // create an list of empty shapes only if there is not static argument
+    if (!spec_.HasArgument(name)) {
+      tls = {};
+    }
+  }
+}
+
+void NumpyReader::Prefetch() {
```

Comment:
I would ad some comment why a custom Prefetch is needed.

Reply:
done, added some comments.